platform/mbed_mktime.c

```diff
@@ -16,14 +16,17 @@
 
 #include "mbed_mktime.h"
 
-/*
- * time constants 
- */
+/* Time constants. */
 #define SECONDS_BY_MINUTES 60
 #define MINUTES_BY_HOUR 60
 #define SECONDS_BY_HOUR (SECONDS_BY_MINUTES * MINUTES_BY_HOUR)
 #define HOURS_BY_DAY 24 
 #define SECONDS_BY_DAY (SECONDS_BY_HOUR * HOURS_BY_DAY)
+#define LAST_VALID_YEAR 206
+
+/* Macros which will be used to determine if we are within valid range. */
+#define EDGE_TIMESTAMP_FULL_LEAP_YEAR_SUPPORT 3220095     // 7th of February 1970 at 06:28:15
+#define EDGE_TIMESTAMP_PARTIAL_LEAP_YEAR_SUPPORT 3133695  // 6th of February 1970 at 06:28:15
```

Comment:
Why was this date chosen? Also, why is it different for partial vs full?

Reply:
This is a shortcut to check if calendar date provided to _rtc_mktime() function is within valid range. This is used to check upper limit. Function first only checks if the year range is valid. Then it calculates elapsed time since the beginning of the specified year. At this point we can check if provided date is within the range using these macros. If specified year is 2106 (last valid) and seconds since the beginning of the year are greater than edge timestamp, then we are out of range.
Different values are used for these two types of RTC devices. On devices which support full leap year detection last valid point of time is the 7th of February 2106 at 06:28:15 and on devices which have only 4 year leap year detection last valid date is 6th of February 2106 at 06:28:15 (at this point we are one day off). Both dates correspond to 0xffff_ffff number of seconds since the start of unix epoch.